src/backend/distributed/planner/multi_logical_optimizer.c

```diff
@@ -1959,6 +2021,145 @@ WorkerExtendedOpNode(MultiExtendedOp *originalOpNode,
 
 
 /*
+ * HasNonPartitionColumnDistinctAgg returns true if target entry or having qualifier
+ * has non-partition column reference in aggregate definition. Note that, it only
+ * checks aggs subfield of Aggref, it does not check FILTER or SORT clauses.
+ */
+static bool
+HasNonPartitionColumnDistinctAgg(List *targetEntryList, Node *havingQual,
+								 List *tableNodeList)
+{
+	List *targetVarList = pull_var_clause((Node *) targetEntryList,
+										  PVC_INCLUDE_AGGREGATES);
+	List *havingVarList = pull_var_clause((Node *) havingQual, PVC_INCLUDE_AGGREGATES);
+	List *aggregateCheckList = list_concat(targetVarList, havingVarList);
+	ListCell *aggregateCheckCell = NULL;
+	foreach(aggregateCheckCell, aggregateCheckList)
+	{
+		Node *targetNode = lfirst(aggregateCheckCell);
+		Aggref *targetAgg = NULL;
+		List *varList = NIL;
+		ListCell *varCell = NULL;
+		bool isPartitionColumn = false;
+
+		if (IsA(targetNode, Var))
+		{
+			continue;
+		}
+
+		Assert(IsA(targetNode, Aggref));
+		targetAgg = (Aggref *) targetNode;
+		if (targetAgg->aggdistinct == NIL)
+		{
+			continue;
+		}
+
+		varList = pull_var_clause_default((Node *) targetAgg->args);
+
+		if (varList == NIL)
+		{
+			continue;
+		}
+
+		foreach(varCell, varList)
+		{
+			Node *targetVar = (Node *) lfirst(varCell);
+			Assert(IsA(targetVar, Var));
+
+			isPartitionColumn = ExpressionContainsPartitionColumn(
+				targetVar, tableNodeList);
+
+			if (!isPartitionColumn)
+			{
+				return true;
+			}
+		}
+	}
+
+	return false;
+}
+
+
+/*
+ * ShouldPullDistinctColumn returns true if distinct aggregate should pull
+ * individual columns from worker to master and evaluate aggregate operation
+ * at master.
+ *
+ * Pull cases are:
+ * - repartition subqueries
+ * - query has count distinct on a non-partition column on at least one target
+ * - count distinct is on a non-partition column and query is not
+ *   grouped on partition column
+ */
+static bool
+ShouldPullDistinctColumn(bool groupedByDisjointPartitionColumn,
```

Comment:
I find a bit harder to follow this code path, based on on your comment what about refactoring like this (also reordered parameters):

```c
static bool
ShouldPullDistinctColumn(Node *expression, List *tableNodeList,
						bool repartitionSubquery,
						bool groupedByDisjointPartitionColumn,
						bool hasNonPartitionColumnDistinctAgg)
{
	if (repartitionSubquery)
	{
		return true;
	}

	if (!groupedByDisjointPartitionColumn)
	{
		bool hasPartitionColumn = false;

		if (hasNonPartitionColumnDistinctAgg)
		{
			return true;
		}

		hasPartitionColumn = ExpressionContainsPartitionColumn(expression, tableNodeList);
		if (!hasPartitionColumn)
		{
			return true;
		}
	}

	return false;
}
```

Reply:
Also, is this correct? Can we even simply this?

Reply:
looks okay to me